### PR TITLE
Remove SubselectRewriter

### PR DESCRIPTION
**Pull request by mfussenegger**
The way the `LogicalPlanner` works we end up getting similar or the same
execution plans without having the `SubselectRewriter` doing its
collapsing logic.

There are two reasons to remove the `SubselectRewriter`:

 - It currently relies a lot on the `QuerySpec` and the ability to
 mutate it freely. This mutation is problematic because we've no means
 in place to safe-guard invariants - for example that the fields of a
 relation must match the outputs of a `QuerySpec`.

 - The analyzer shouldn't be doing optimizations, even if they're always
 applicable. We've already a infrastructure in the Planner to optimize
 plans and should use that instead.

This commit introduces a performance regression affecting JOINS
involving subqueries as the `WHERE` clause push-down, no longer happens
in all cases in the same way. There will be a follow-up to re-add this
using the `LogicalPlan#tryOptimize` functionality.

### CODE REVIEW

**Comment by mfussenegger** [comment]
> Should we do this optimization in the LogicalPlanner too? 

Which optimization?

See the description:

> This commit introduces a performance regression affecting JOINS
involving subqueries as the WHERE clause push-down, no longer happens
in all cases in the same way. There will be a follow-up to re-add this
using the LogicalPlan#tryOptimize functionality.

Opened up https://github.com/crate/crate/pull/7080 already

**Comment by matriv** [comment]
Missed to paste... https://github.com/crate/crate/pull/7076/files#diff-b28d3d3d0ecba6743d1d2c732c37a91cL53 The limit collapsing.

File: sql/src/test/java/io/crate/testing/SQLPrinter.java [hunk 1]
```diff
@@ -49,6 +50,12 @@ public static String print(Object o) {
             return print((Collection<Symbol>) o);
         } else if (o == null) {
             return "null";
+        } else if (o instanceof QueryClause) {
+            if (((QueryClause) o).hasQuery()) {
```
Comment (matriv) [inline sql/src/test/java/io/crate/testing/SQLPrinter.java]: When we use isSQL() with that we cannot distinguish between MATCH_ALL and NO_MATCH.
I guess we would need something like:
```
QueryClause qc = (QueryClause) o;
if (qc.hasQuery()) {
    return print(((QueryClause) o).query());
} else if (qc.noMatch()) {
    return "NO_MATCH";
} else {
    return "MATCH_ALL";
}
```
or would it break something else?

Comment (mfussenegger) [reply sql/src/test/java/io/crate/testing/SQLPrinter.java]: @matriv pushed a fixup for this

Comment (matriv) [reply sql/src/test/java/io/crate/testing/SQLPrinter.java]: Thanks! acked!

**Pull request merged**